src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs

```diff
@@ -281,7 +281,14 @@ internal int ReadCore(Span<byte> buffer)
                     int n = _stream.Read(_buffer, 0, _buffer.Length);
                     if (n <= 0)
                     {
-                        break;
+                        if (!_inflater.Finished() && _inflater.NonZeroInput() && _inflater.AvailableOutput > 0)
+                        {
+                            ThrowGenericInvalidData();
+                        }
+                        else
```

Comment:
is the 'else' needed here - eg codegen improves because it does not realize that the throw method never returns?

Reply:
I'm not sure how to properly check this but I'll look into it.

Reply:
If I checked things correctly (by codegen you meant IL and not x86, right?), this is the result. With the `else`
```
      // [286 28 - 286 54]
      IL_00bc: call         void System.IO.Compression.DeflateStream::ThrowGenericInvalidData()
      IL_00c1: nop          

      // [287 25 - 287 26]
      IL_00c2: nop          

      IL_00c3: br.s         IL_00c8

      // [289 25 - 289 26]
      IL_00c5: nop          

      // [290 29 - 290 35]
      IL_00c6: br.s         IL_0120

      // [292 21 - 292 22]
      IL_00c8: nop          

      IL_00c9: br.s         IL_00fc
```

Without the `else`
```
      // [286 28 - 286 54]
      IL_00bc: call         void System.IO.Compression.DeflateStream::ThrowGenericInvalidData()
      IL_00c1: nop          

      // [287 25 - 287 26]
      IL_00c2: nop          

      // [290 29 - 290 35]
      IL_00c3: br.s         IL_011a
```

So it seems you are correct! I updated the PR accordingly.

Reply:
By codegen we generally mean assembly code. Usually the quickest way to verify that is with sharplab.io. Not sure how much the IL correlates. But it makes sense to change as you suggest.